src/constraints.rs

```diff
@@ -0,0 +1,536 @@
+where
+    E: PairingEngine,
+    P: PairingVar<E, E::Fq>,
+{
+}
+
+impl<E, P> AllocVar<PreparedVerifyingKey<E>, E::Fq> for PreparedVerifyingKeyVar<E, P>
+where
+    E: PairingEngine,
+    P: PairingVar<E>,
+{
+    #[tracing::instrument(target = "r1cs", skip(cs, f))]
+    fn new_variable<T: Borrow<PreparedVerifyingKey<E>>>(
+        cs: impl Into<Namespace<E::Fq>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        f().and_then(|pvk| {
+            let pvk = pvk.borrow();
+            let alpha_g1_beta_g2 = P::GTVar::new_variable(
+                ark_relations::ns!(cs, "alpha_g1_beta_g2"),
+                || Ok(pvk.alpha_g1_beta_g2.clone()),
+                mode,
+            )?;
+
+            let gamma_g2_neg_pc = P::G2PreparedVar::new_variable(
+                ark_relations::ns!(cs, "gamma_g2_neg_pc"),
+                || Ok(pvk.gamma_g2_neg_pc.clone()),
+                mode,
+            )?;
+
+            let delta_g2_neg_pc = P::G2PreparedVar::new_variable(
+                ark_relations::ns!(cs, "delta_g2_neg_pc"),
+                || Ok(pvk.delta_g2_neg_pc.clone()),
+                mode,
+            )?;
+
+            let gamma_abc_g1 = Vec::new_variable(
+                ark_relations::ns!(cs, "gamma_abc_g1"),
+                || Ok(pvk.vk.gamma_abc_g1.clone()),
+                mode,
+            )?;
+
+            Ok(Self {
+                alpha_g1_beta_g2,
+                gamma_g2_neg_pc,
+                delta_g2_neg_pc,
+                gamma_abc_g1,
+            })
+        })
+    }
+}
+
+impl<E, P> AllocVar<VerifyingKey<E>, E::Fq> for VerifyingKeyVar<E, P>
+where
+    E: PairingEngine,
+
+    P: PairingVar<E>,
+{
+    #[tracing::instrument(target = "r1cs", skip(cs, f))]
+    fn new_variable<T: Borrow<VerifyingKey<E>>>(
+        cs: impl Into<Namespace<E::Fq>>,
+        f: impl FnOnce() -> Result<T, SynthesisError>,
+        mode: AllocationMode,
+    ) -> Result<Self, SynthesisError> {
+        let ns = cs.into();
+        let cs = ns.cs();
+
+        f().and_then(|vk| {
+            let VerifyingKey {
+                alpha_g1,
+                beta_g2,
+                gamma_g2,
+                delta_g2,
+                gamma_abc_g1,
+            } = vk.borrow().clone();
+            let alpha_g1 =
+                P::G1Var::new_variable(ark_relations::ns!(cs, "alpha_g1"), || Ok(alpha_g1), mode)?;
+            let beta_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "beta_g2"), || Ok(beta_g2), mode)?;
+            let gamma_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "gamma_g2"), || Ok(gamma_g2), mode)?;
+            let delta_g2 =
+                P::G2Var::new_variable(ark_relations::ns!(cs, "delta_g2"), || Ok(delta_g2), mode)?;
+
+            let gamma_abc_g1 = Vec::new_variable(cs.clone(), || Ok(gamma_abc_g1), mode)?;
```

Comment:
I feel like the newline should be after this line. Also, is there a reason this doesn't get a namespace like the others do?